### PR TITLE
Update connection management in util and rxm provider

### PR DESCRIPTION
**Pull request by a-ilango**
- common: Add a function to compare addresses	
- util: Augment cmap to handle additional states.
- rxm: Pass handle key as context for connection handling to avoid NULL
  pointer dereferences.
- rxm: Make the listener thread per endpoint to simplify processing connection
  events

### CODE REVIEW

**Comment by shefty** [comment]
@a-ilango - See OS X failure: fi_rdm_pingpong -I 5 -p "ofi-rxm"  result: Fail

File: include/fi_util.h [hunk 1]
```diff
@@ -376,12 +373,18 @@ int ofi_av_get_index(struct util_av *av, const void *addr);
...
+// TODO explore replacing this with a simple connection hash map that is common
+// for both AV and RX only connections.
 #define UTIL_CMAP_IDX_BITS 48
 
 enum util_cmap_state {
 	CMAP_UNSPEC,
-	CMAP_CONNECTING,
-	CMAP_CONNECTED
+	CMAP_CONN_INIT,
+	CMAP_CONNECT,
```
Comment (shefty) [inline include/fi_util.h]: The CONNECT state is unclear relative to CONNREQ, ACCEPT, CONNECTED,

File: include/fi_util.h [hunk 2]
```diff
@@ -418,23 +433,26 @@ struct util_cmap {
 
 	struct ofi_key_idx key_idx;
 
+	uint8_t closing;
```
Comment (shefty) [inline include/fi_util.h]: Can we expand/use the cmap_state for this, rather than maintaining another state variable?

File: prov/rxm/src/rxm_conn.c [hunk 3]
```diff
@@ -190,49 +172,60 @@ void *rxm_msg_listener(void *arg)
...
-			FI_WARN(&rxm_prov, FI_LOG_FABRIC, "Unknown event: %u\n", event);
+			FI_WARN(&rxm_prov, FI_LOG_FABRIC,
+				"Unknown event: %u\n", event);
 		}
 		continue;
 err:
-		FI_WARN(&rxm_prov, FI_LOG_FABRIC,
-				"Received size (%d) not matching expected (%d)\n", rd, len);
+		FI_WARN(&rxm_prov, FI_LOG_FABRIC, "Received size (%d) not "
+			"matching expected (%d)\n", rd, len);
```
Comment (shefty) [inline prov/rxm/src/rxm_conn.c]: This code doesn't seem reachable.

File: prov/rxm/src/rxm_ep.c [hunk 4]
```diff
@@ -1068,34 +1100,58 @@ static int rxm_ep_msg_res_open(struct fi_info *rxm_fi_info,
...
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+				"Unable to bind msg PEP to msg EQ\n");
+		goto err4;
+	}
+	ret = fi_listen(rxm_ep->msg_pep);
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+				"Unable to set msg PEP to listen state\n");
+		goto err4;
+	}
```
Comment (shefty) [inline prov/rxm/src/rxm_ep.c]: I would move the listen handling portion into another function to shorten this one and group that functionality together.  Only start the listen thread after everything else has been setup.

File: include/fi_util.h (hunk 2 above)
Comment (a-ilango) [inline include/fi_util.h]: This variable is used to signal that the whole of cmap is about to be closed / freed. The cmap_state is for individual handles / connection in the cmap.

File: include/fi_util.h (hunk 1 above)
Comment (a-ilango) [inline include/fi_util.h]: The name could be left as CONNECTING if that's more clear.

File: prov/rxm/src/rxm_conn.c (hunk 3 above)
Comment (a-ilango) [inline prov/rxm/src/rxm_conn.c]: This code would be reached as part of error handling inside some of the switch cases.

File: prov/rxm/src/rxm_ep.c (hunk 4 above)
Comment (a-ilango) [inline prov/rxm/src/rxm_ep.c]: 👍

File: include/fi_util.h (hunk 1 above)
Comment (shefty) [inline include/fi_util.h]: I'm not sure that's clearer.  The states below: CONNREQ and ACCEPT are also part of "connecting".  Does this state mean CONNREQ_SENT?
On this note, without looking at the code, it's not clear why there's a CONN_INIT state.  That state seems like it be immediately transitioned into actively connecting.

File: include/fi_util.h (hunk 2 above)
Comment (shefty) [inline include/fi_util.h]: A state that indicates a structure is about to be freed isn't useful unless there's some other reference counting occurring. I don't see this field being used.

File: prov/rxm/src/rxm_conn.c (hunk 3 above)
Comment (shefty) [inline prov/rxm/src/rxm_conn.c]: I still can't see how it can be reached.  I don't see a 'goto err', and immediately before this label is a continue call.  Maybe it's reached by code not shown in this patch...

Comment (a-ilango) [inline prov/rxm/src/rxm_conn.c]: Yes, it's not part of this patch. There is just one instance of goto err. I guess this code can be moved there instead.

File: include/fi_util.h (hunk 1 above)
Comment (a-ilango) [inline include/fi_util.h]: Yes this state would correspond to CONNREQ_SENT. I think the CONN_INIT state could be done away with. It was initially added on the first iteration of this change when it was needed back then.

Comment (shefty) [inline include/fi_util.h]: Okay, then I would rename CONNREQ to CONNREQ_RECV.  I was interpretting that as the sent state.

File: include/fi_util.h [hunk 5]
```diff
@@ -376,12 +373,17 @@ int ofi_av_get_index(struct util_av *av, const void *addr);
...
 
+// TODO explore replacing this with a simple connection hash map that is common
+// for both AV and RX only connections.
 #define UTIL_CMAP_IDX_BITS 48
 
 enum util_cmap_state {
 	CMAP_UNSPEC,
-	CMAP_CONNECTING,
-	CMAP_CONNECTED
+	CMAP_CONNECT,
```
Comment (shefty) [inline include/fi_util.h]: Can this be renamed to CONNREQ_SENT?

File: include/fi_util.h [hunk 6]
```diff
@@ -400,13 +402,25 @@ struct util_cmap_handle {
...
+typedef void (*ofi_cmap_handle_func)(struct util_cmap_handle *handle);
+typedef int (*ofi_cmap_connect_func)(struct util_ep *cmap,
+				     struct util_cmap_handle *handle,
+				     fi_addr_t fi_addr);
+
+struct util_cmap_attr {
+	void 				*name;
+	ofi_cmap_alloc_handle_func 	alloc;
+	ofi_cmap_handle_func 		close;
+	ofi_cmap_handle_func 		free;
```
Comment (shefty) [inline include/fi_util.h]: Just looking at the names, it looks odd having both close and free calls.  Can these be combined?  Does close mean disconnect?

File: include/fi_util.h (hunk 5 above)
Comment (a-ilango) [inline include/fi_util.h]: Sure.

File: include/fi_util.h (hunk 6 above)
Comment (a-ilango) [inline include/fi_util.h]: The close call closes the connection and associated resources. The free function frees the memory allocated for handle. We have to wait for the shutdown event to free the allocated memory to avoid referencing freed memory. That's why they can't be combined.

Comment (shefty) [inline include/fi_util.h]: Can we rename close to disconnect then?  That would make it easier to associate with the connect call below.

Comment (a-ilango) [inline include/fi_util.h]: Sorry. I should have been more clear. The close call closes the resources which automatically disconnects the connection. We don't explicitly disconnect.

**Comment by a-ilango** [comment]
@shefty I've updated the cmap to accept an event handler and signal function so that it has the control to start / stop the event handler / listener thread. This avoids certain race conditions and the OS X segfault issue seems to have gone now. I'm not entirely sure if this update fixed that. :)
Also disabled some assertions in certain error cases. Those assertions are being hit only when run over sockets provider.
